Fix Pay via Paypal block: currency symbol appears after the price in editor

mreishus opened the pull request:
fixes #11897

#### Changes proposed in this Pull Request:

* Simple payments block now uses `Intl.NumberFormat()` to format currency display, if available

#### Jetpack product discussion

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* On master branch, create a new post.
* Add a 'Pay via Paypal' block and fill out the information.  Save as draft.
* Unfocus the block.  In preview mode, instead of "$5.00", it will say "5.00 $".
* Switch to this branch and reload the post.
* The block should upgrade correctly and now display $5.00.
* Other currencies and locales should be supported as well.

![2020-10-06_14-25](https://user-images.githubusercontent.com/937354/95254008-e15f7700-07e4-11eb-870c-44ff5bd5c951.png)

( To test on my sandbox, I used `yarn build-extensions --watch` and then rsynced the `./_inc/blocks` folder over).

#### Proposed changelog entry for your changes:

* Simple payments block now uses `Intl.NumberFormat()` to format currency display, if available

mreishus committed:
[not verified] Pay via Paypal block: currency symbol appears after the price in editor

Issue: https://github.com/Automattic/jetpack/issues/11897

Approach: Use Intl.NumberFormat (pretty defensively) to format
currencies with symbol.

mreishus committed:
[not verified] Simple-Payments: Add deprecated v2 with old version of formatPrice()

The previous commit changed formatPrice() to be smarter.
This deprecated version is the same as current, but replaces calls to
formatPrice() with formatPriceFallback().

mreishus commented on extensions/blocks/simple-payments/utils.js:
```diff
@@ -21,9 +22,50 @@ export const decimalPlaces = number => {
+		return navigator.languages[ 0 ];
+	}
+	return navigator.userLanguage || navigator.language || navigator.browserLanguage || 'en';
+};
```
I grabbed this after looking through https://stackoverflow.com/questions/673905/best-way-to-determine-users-locale-within-browser .

simison replied:
Wouldn't you need it in site's locale, rather than user's browser? 🤔 I might be missing some context here tho. :-)

mreishus replied:
Yeah.. we're trying to find the best way to get the site's locale.  I've removed a couple of methods that don't work and the navigator locale is all that's left right now.  I'll add `[WIP]` to the PR to be a bit clearer

obenland commented on extensions/blocks/simple-payments/depecrated/v2/save.js:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Internal dependencies
+ */
+import { formatPriceFallback } from '../..//utils';
```
Found a double slash in case that matters. It probably doesn't

cpapazoglou reviewed (changes requested):
Thanks @mreishus for working on this!
Indeed by applying the patch the euro / dollar sign moved to the left both in editor and the frontend
![](https://cln.sh/1agiRW+)

It seems though, that language discovered is the interface language and not the site language (left inline comment).

Also, have you considered updating this package
https://github.com/Automattic/wp-calypso/blob/2eff0c49dfee1fb27461bd1e11ee4e93d2d83e6e/packages/format-currency/src/index.js#L27

instead of introducing changes that are scoped to only that block?

cpapazoglou commented on extensions/blocks/simple-payments/utils.js:
```diff
@@ -21,9 +22,50 @@ export const decimalPlaces = number => {
+
+// Display prices using Intl.NumberFormat - supported in 95.75% of browsers as of Oct 2020.
+const formatPriceIntlNumberFormat = ( price, currency, withSymbol = true ) => {
+	const { siteLocale } = select( 'core/block-editor' ).getSettings();
```
`const { siteLocale } = select( 'core/block-editor' ).getSettings();` will return the current user interface language and not the current site language. So, when a user (me) has `en` as interface language (very common) but `greek` as site language then the symbol appears on the left whereas it should appear in the right.

![](https://cln.sh/ks7xdO+)

I searched it a bit, I am not sure if `getLocaleSlug` from `i18n-calypso` will return the correct locale
https://github.com/Automattic/wp-calypso/blob/aa011d0e42b1190ead34a407538d21e7e8fe2186/packages/i18n-calypso/src/i18n.js#L314

mreishus replied:
Good catch! I pushed an update that uses getLocalSlug, can you try it out and let me know if it works? I tried updating both my site and wordpress user language to Greek, and it didn't seem to work.  Maybe it's because I'm using the wp-admin edit to test?  This may be difficult..

cpapazoglou replied:
It still returns the `en` version for me too. Seems like 
https://github.com/Automattic/wp-calypso/blob/aa011d0e42b1190ead34a407538d21e7e8fe2186/packages/i18n-calypso/src/i18n.js#L256-L259
it returns the `defaultLocaleSlug`.... That said, I am not sure how we can fix this.

cpapazoglou commented on extensions/blocks/simple-payments/utils.js:
```diff
@@ -21,9 +22,50 @@ export const decimalPlaces = number => {
+
+// Format a price with Intl.NumberFormat if available, but fall back to old method if not.
+export const formatPrice = ( price, currency, withSymbol = true ) => {
+	if ( window.Intl && 'function' === typeof Intl.NumberFormat ) {
+		return formatPriceIntlNumberFormat( price, currency, withSymbol );
+	}
+	return formatPriceFallback( price, currency, withSymbol );
```
Can't this be part of a generic `formatPriceIntlNumberFormat` which will return `formatPriceFallback( price, currency, withSymbol );` if not `window.Intl && 'function' === typeof Intl.NumberFormat`?

mreishus replied:
I'm not sure what you're referring to - is it this? 
* [fix/currency-display c9aa474b6] Inline formatPrice selector function
Personally, I prefer the previous version that has a slightly simpler function at the trade-off of adding one more function.

cpapazoglou replied:
Yes, that is exactly what I meant. 

> Personally, I prefer the previous version that has a slightly simpler function at the trade-off of adding one more function.

From a reading point (while reviewing), I felt it weird but feel free to revert using your version in that case!

mreishus committed:
[not verified] Remove doubleslash

mreishus commented:
> Also, have you considered updating this package
> https://github.com/Automattic/wp-calypso/blob/2eff0c49dfee1fb27461bd1e11ee4e93d2d83e6e/packages/format-currency/src/index.js#L27
> 
> instead of introducing changes that are scoped to only that block?

If we're going to use locale, I don't think `formatCurrency` from `@automattic/format-currency` is a good fit.  It doesn't expect a locale in its API, and I don't think it will be able to pull one from say, calypso state because it's a separate package (would need to be usable by outside consumers).

That said, I'm not sure if we can get the correct locale either, looking into that next..

mreishus commented:
Found an interesting example while testing, in the German locale the Euro sign goes on the right, in English, it goes on the left:

```
console.log(new Intl.NumberFormat('de-DE', { style: 'currency', currency: 'EUR' }).format(12345.67));
12.345,67 €
console.log(new Intl.NumberFormat('en-US', { style: 'currency', currency: 'EUR' }).format(12345.67));
€12,345.67
```

mreishus committed:
[not verified] formatPriceIntlNumberFormat: Try using getLocaleSlug as experiment

mreishus committed:
[not verified] Inline formatPrice selector function

mreishus committed:
[not verified] Revert "formatPriceIntlNumberFormat: Try using getLocaleSlug as experiment"

This reverts commit 92c48efe3bffafd9fb3174e376ed8f7bdfa55458.

- This didn't seem to work

mreishus commented:
It looks like jetpack removed all references to `i18n-calypso` and uses `@wordpress/i18n` for everything. (See: https://github.com/Automattic/jetpack/issues/16481)

Since `@wordpress/i18n` knows how to translate into the user language ( `__("This is a test")` ), it has to know the user language.  So, can we get the current user language from it?  I see a `setLocaleData` but not a `getLocale` from https://github.com/WordPress/gutenberg/tree/master/packages/i18n.

mreishus commented:
We might be able to get it by translating `html_lang_attribute` with `__()`.  I found this in some translation files.

![2020-10-07_12-29](https://user-images.githubusercontent.com/937354/95366239-c8b09900-0898-11eb-8fb6-e914aef4599e.png)
![2020-10-07_12-29_1](https://user-images.githubusercontent.com/937354/95366246-cbab8980-0898-11eb-8c51-07450d869b0b.png)

mreishus commented:
Not having much luck with `__`.  I'm not able to get it to translate anything at all; even after changing my wordpress interface language, site language in wp-admin, and browser language to French, I can't get a translation that's elsewhere in the code to work:

```
	const backup = __( 'Daily Backups', 'jetpack' );
	console.log( backup );
```

mreishus committed:
[not verified] Remove broken locale code

mreishus commented:
I removed all of the locale detection code, since it wasn't working.  The function that gets it from the browser settings is there and seems to work.  That might be good?

![2020-10-07_14-02](https://user-images.githubusercontent.com/937354/95376162-47600300-08a6-11eb-8c53-113e5ec34271.png)
![2020-10-07_14-04](https://user-images.githubusercontent.com/937354/95376166-49c25d00-08a6-11eb-950e-8dec74239a90.png)

Next steps(?):
* Ask Jetpack crew if there's a way to get the current locale
* If not, reconsider adding this code to the '@automattic/format-currency' package

simison commented on extensions/blocks/simple-payments/index.js:
```diff
@@ -189,5 +190,5 @@ export const settings = {
 	},
 
-	deprecated: [ deprecatedV1 ],
+	deprecated: [ deprecatedV1, deprecatedV2 ],
```
Not blocking this PR or anything, but I was always wondering if it would work to async-load these deprecations. They're basically doubling/tripling the block JS+CSS size each time.

cc @sgomes for thoughts

sgomes replied:
If at all possible, I would heartily recommend that, yes. Looking at the code, however, it's not immediately obvious how to do that easily.

Perhaps making `deprecated` an async function to an array, rather than a simple array?

```js
{
  deprecated: async () => {
    const deprecatedV1 = await import( './depecrated/v1' );
    const deprecatedV2 = await import( './depecrated/v2' );
    return [ deprecatedV1, deprecatedV2 ];
  },
}
```

Other code that makes use of these properties would then have to be updated accordingly, and made to be async-safe.

Fixing the `depecrated` typo would also be nice 😄

simison replied:
Yeah, so maybe this is something to add support for in core first. cc @mtias @youknowriad

jeherve replied:
Conversation to be continued in #17467.

cpapazoglou commented:
> I removed all of the locale detection code, since it wasn't working. The function that gets it from the browser settings is there and seems to work. That might be good?
> 
> ![2020-10-07_14-02](https://user-images.githubusercontent.com/937354/95376162-47600300-08a6-11eb-8c53-113e5ec34271.png)
> ![2020-10-07_14-04](https://user-images.githubusercontent.com/937354/95376166-49c25d00-08a6-11eb-950e-8dec74239a90.png)
> 
> Next steps(?):
> 
> * Ask Jetpack crew if there's a way to get the current locale
> * If not, reconsider adding this code to the '@automattic/format-currency' package

If getting the current site locale seems impossible, lets go with the browser locale which shouldn't be that bad from a visitor perspective and we can iterate later.

It seems also that we should have a holistic solution for currency symbols across all blocks that show prices.

mreishus committed:
[not verified] Return to using interface locale

We don't know of a way to get the site locale.

Jeremy offered:
const { l10n: { locale } } = __experimentalGetSettings();

In my testing, this also returns the interface locale.

mreishus committed:
[not verified] getNavigatorLanguage uses global.window to access navigator

mreishus committed:
Merge branch 'master' into fix/currency-display

mreishus committed:
Merge branch 'master' into fix/currency-display

mreishus commented:
@cpapazoglou : Do you have an opinion on using the browser language vs `const { siteLocale } = select( 'core/block-editor' ).getSettings();` which gets the wp.com interface language?  Which is better?

cpapazoglou commented:
> @cpapazoglou : Do you have an opinion on using the browser language vs `const { siteLocale } = select( 'core/block-editor' ).getSettings();` which gets the wp.com interface language? Which is better?

I would go with `const { siteLocale } = select( 'core/block-editor' ).getSettings();` which is something users can set to our application and we don't rely to browser settings (which in some cases users may not have access to also).

mreishus commented:
Ok, summarizing the issues from reviews so far:

- **How to load the WordPress site's locale?**
After searching around, we couldn't find a way to do it.  The code currently uses the Wordpress interface locale, and if that isn't valid, then falls back to the browser's locale.
- **Async load deprecations**
I split this off into its own issue: https://github.com/Automattic/jetpack/issues/17467
- **Apply the fix to all blocks that show currencies?**
Calypso does have an `@automattic/format-currency` package, which contains a `formatCurrency()` function.  That function doesn't expect its callers to pass in a locale.  If we were to modify `formatCurrency()` to use `Intl.NumberFormat()`, all of its call sites would have to be modified to find a locale to pass, and the scope of this change would increase considerably.  
Given the number of user issues inside the parent issue, I propose moving forward with the limited scope fix and put modifying `@automattic/format-currency` into another issue.

cpapazoglou commented:
> Given the number of user issues inside the parent issue, I propose moving forward with the limited scope fix and put modifying @automattic/format-currency into another issue.

Sounds good!

cpapazoglou reviewed (approved):
Looks and works great! Thanks for making the changes!

jeherve commented on extensions/blocks/simple-payments/index.js:
```diff
@@ -14,6 +14,7 @@ import { getIconColor } from '../../shared/block-icons';
 import edit from './edit';
 import save from './save';
 import deprecatedV1 from './depecrated/v1';
+import deprecatedV2 from './depecrated/v2';
```
Maybe we can take this opportunity to:

1. Rename `depecrated/` into `deprecated` to fix this typo.
2. Have a `deprecated/index.js` where we collect all deprecations, a bit like here:
https://github.com/Automattic/jetpack/blob/c3774de77e1c255b7e5e447cbc246dc21daca7e4/extensions/blocks/eventbrite/deprecated/index.js#L2-L7

jeherve reviewed (approved):
This looks good to me. I have some feedback below, but that's a minor nitpick, can be done in another PR.

mreishus committed:
Merge branch 'master' into fix/currency-display

mreishus commented:
Github lists one of the required checks as failing: "wpcom — Tests failed: D50729-code".  However, looking over at D50729-code, it shows all the checks in the buildable passing.  Any insight into this @jeherve ?

AtrumGeost commented:
Hi folks!

Thanks for working on this :) Do you know when these changes will be reflected on WordPress.com?

mreishus commented:
Hi @AtrumGeost

> Thanks for working on this :) Do you know when these changes will be reflected on WordPress.com?

Jetpack is released on the first Tuesday of each month, so it should be November 3rd (someone please correct me if I'm wrong).

simison commented:
@mreishus are you waiting for the release day to ship .com side? D50729-code

mreishus commented:
> @mreishus are you waiting for the release day to ship .com side? [D50729-code](https://href.li/?D50729-code)

That does work, but I was a bit worried about Simple -> Atomic transfers occurring after that was pushed.  There are invisible block validation errors, which might not matter?  But I was thinking about the case where someone pays extra to get a business or ecommerce plan, finds their paypal block is degraded as a part of the transfer, and then there's nothing we can do about it.  Any thoughts on that?

jeherve commented:
> Jetpack is released on the first Tuesday of each month, so it should be November 3rd

That would normally be the case, but we're exceptionally releasing a week later this month, to avoid releasing on Election Day in the US.

> I was thinking about the case where someone pays extra to get a business or ecommerce plan, finds their paypal block is degraded as a part of the transfer, and then there's nothing we can do about it.

Yes, I think it's fine to wait until release day to push that commit. That's what we've been doing so far for all commits that change the way blocks are saved, to avoid the problems you highlighted.

jeherve commented:
r216551-wpcom
